File: packages/python/minions_openclaw/snapshot_manager.py
"""Snapshot manager."""
from __future__ import annotations
import json
from pathlib import Path
from typing import Any, Dict, List, Optional

from minions import Minion, Relation, create_minion, generate_id, now
from .types import openclaw_snapshot_type

DATA_DIR = Path.home() / '.openclaw-manager'
DATA_FILE = DATA_DIR / 'data.json'


def _read_storage() -> Dict[str, Any]:
    DATA_DIR.mkdir(parents=True, exist_ok=True)
    try:
        return json.loads(DATA_FILE.read_text())
    except (FileNotFoundError, json.JSONDecodeError):
        return {'minions': [], 'relations': []}


def _write_storage(data: Dict[str, Any]) -> None:
    DATA_DIR.mkdir(parents=True, exist_ok=True)
    DATA_FILE.write_text(json.dumps(data, indent=2))


class SnapshotManager:
    def capture_snapshot(self, instance_id: str, gateway_data: Dict[str, Any]) -> Minion:
        storage = _read_storage()
        minion, _ = create_minion(
            {
                "title": f"Snapshot {now()}",
                "fields": {
                    'instanceId': instance_id,
                    'capturedAt': now(),
                    'config': json.dumps(gateway_data.get('config', {})),
                    'agentCount': len(gateway_data.get('agents', [])),
                    'channelCount': len(gateway_data.get('channels', [])),
                    'modelCount': len(gateway_data.get('models', [])),
                }
            },
            openclaw_snapshot_type
        )
        minion_dict = {
            'id': minion.id,
            'title': minion.title,
            'minionTypeId': minion.minion_type_id,
            'fields': minion.fields,
            'createdAt': minion.created_at,
            'updatedAt': minion.updated_at,
            'tags': minion.tags,
            'status': minion.status,
            'priority': minion.priority,
            'description': minion.description,
        }
        storage['minions'].append(minion_dict)
        storage['relations'].append({
            'id': generate_id(),
            'sourceId': instance_id,
            'targetId': minion.id,
            'type': 'parent_of',
            'createdAt': now(),
            'metadata': {},
        })
        _write_storage(storage)
        return minion

    def list_snapshots(self, instance_id: str) -> List[Dict[str, Any]]:
        storage = _read_storage()
        snapshot_ids = {
            r['targetId'] for r in storage['relations']
            if r.get('sourceId') == instance_id and r.get('type') == 'parent_of'
        }
        return [
            m for m in storage['minions']
            if m.get('id') in snapshot_ids
            and m.get('minionTypeId') == openclaw_snapshot_type.id
            and not m.get('deletedAt')
        ]

    def get_history(self, instance_id: str) -> List[Dict[str, Any]]:
        """Return snapshots for instance ordered newest → oldest via follows chain."""
        storage = _read_storage()
        snapshot_ids = {
            r['targetId'] for r in storage['relations']
            if r.get('sourceId') == instance_id and r.get('type') == 'parent_of'
        }
        snapshots = [
            m for m in storage['minions']
            if m.get('id') in snapshot_ids
            and m.get('minionTypeId') == openclaw_snapshot_type.id
            and not m.get('deletedAt')
        ]

        # Build follows map: snapshot_id → previous_snapshot_id
        follows_map: Dict[str, str] = {}
        for r in storage['relations']:
            if r.get('type') == 'follows' and r.get('sourceId') in snapshot_ids:
                follows_map[r['sourceId']] = r['targetId']

        targets = set(follows_map.values())
        head = next((s for s in snapshots if s['id'] not in targets), None)
        if not head:
            return sorted(snapshots, key=lambda m: m.get('createdAt', ''), reverse=True)

        by_id = {s['id']: s for s in snapshots}
        ordered: List[Dict[str, Any]] = []
        current: Optional[Dict[str, Any]] = head
        while current:
            ordered.append(current)
            next_id = follows_map.get(current['id'])
            current = by_id.get(next_id) if next_id else None
        return ordered

    def compare(self, snapshot_id1: str, snapshot_id2: str) -> Dict[str, Dict[str, Any]]:
        """Load two snapshots by ID and return their diff."""
        storage = _read_storage()
        by_id = {m['id']: m for m in storage['minions']}
        raw_a = by_id.get(snapshot_id1)
        raw_b = by_id.get(snapshot_id2)
        if not raw_a:
            raise ValueError(f"Snapshot not found: {snapshot_id1}")
        if not raw_b:
            raise ValueError(f"Snapshot not found: {snapshot_id2}")

        def _to_minion(d: Dict[str, Any]) -> Minion:
            return Minion(
                id=d['id'],
                title=d.get('title', ''),
                minion_type_id=d.get('minionTypeId', ''),
                fields=d.get('fields', {}),
                created_at=d.get('createdAt', ''),
                updated_at=d.get('updatedAt', ''),
                tags=d.get('tags', []),
                status=d.get('status', 'active'),
                priority=d.get('priority', 'medium'),
                description=d.get('description', ''),
            )

        return self.diff_snapshots(_to_minion(raw_a), _to_minion(raw_b))

    def diff_snapshots(self, a: Minion, b: Minion) -> Dict[str, Dict[str, Any]]:
        """Compare two snapshots and return a dict of changed fields.

        Args:
            a: The first (baseline) snapshot minion.
            b: The second (comparison) snapshot minion.

        Returns:
            Dict mapping field name to {'from': old_value, 'to': new_value}
            for each field that differs between the two snapshots.
        """
        import json as _json
        diff: Dict[str, Dict[str, Any]] = {}
        all_keys = set(list(a.fields.keys()) + list(b.fields.keys()))
        for key in all_keys:
            va = a.fields.get(key)
            vb = b.fields.get(key)
            if _json.dumps(va, sort_keys=True, default=str) != _json.dumps(vb, sort_keys=True, default=str):
                diff[key] = {'from': va, 'to': vb}
        return diff

File: packages/core/src/SnapshotManager.ts
import { createMinion, generateId } from 'minions-sdk';
import type { Minion, Relation } from 'minions-sdk';
import { openclawSnapshotType } from './types.js';
import { promises as fs } from 'fs';
import { homedir } from 'os';
import { join } from 'path';

const DATA_DIR = join(homedir(), '.openclaw-manager');
const DATA_FILE = join(DATA_DIR, 'data.json');

interface StorageData {
  minions: Minion[];
  relations: Relation[];
}

async function readStorage(): Promise<StorageData> {
  try {
    await fs.mkdir(DATA_DIR, { recursive: true });
    const raw = await fs.readFile(DATA_FILE, 'utf-8');
    return JSON.parse(raw) as StorageData;
  } catch {
    return { minions: [], relations: [] };
  }
}

async function writeStorage(data: StorageData): Promise<void> {
  await fs.mkdir(DATA_DIR, { recursive: true });
  await fs.writeFile(DATA_FILE, JSON.stringify(data, null, 2), 'utf-8');
}

export interface GatewaySnapshot {
  agents: unknown[];
  channels: unknown[];
  models: unknown[];
  config: Record<string, unknown>;
}

export class SnapshotManager {
  async captureSnapshot(instanceId: string, gatewayData: GatewaySnapshot): Promise<Minion> {
    const storage = await readStorage();

    const previousSnapshots = storage.minions
      .filter(m => m.minionTypeId === openclawSnapshotType.id && m.deletedAt == null)
      .filter(m => storage.relations.some(r => r.sourceId === instanceId && r.targetId === m.id))
      .sort((a, b) => new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime());

    const { minion } = createMinion(
      {
        title: `Snapshot ${new Date().toISOString()}`,
        fields: {
          instanceId,
          capturedAt: new Date().toISOString(),
          config: JSON.stringify(gatewayData.config),
          agentCount: gatewayData.agents.length,
          channelCount: gatewayData.channels.length,
          modelCount: gatewayData.models.length,
        },
        tags: [],
      },
      openclawSnapshotType
    );

    storage.minions.push(minion);

    storage.relations.push({
      id: generateId(),
      sourceId: instanceId,
      targetId: minion.id,
      type: 'parent_of',
      createdAt: new Date().toISOString(),
      metadata: {},
    });

    if (previousSnapshots.length > 0) {
      storage.relations.push({
        id: generateId(),
        sourceId: minion.id,
        targetId: previousSnapshots[0].id,
        type: 'follows',
        createdAt: new Date().toISOString(),
        metadata: {},
      });
    }

    await writeStorage(storage);
    return minion;
  }

  async listSnapshots(instanceId: string): Promise<Minion[]> {
    const storage = await readStorage();
    const snapshotIds = storage.relations
      .filter(r => r.sourceId === instanceId && r.type === 'parent_of')
      .map(r => r.targetId);
    return storage.minions
      .filter(m => snapshotIds.includes(m.id) && m.minionTypeId === openclawSnapshotType.id && (m.deletedAt === undefined || m.deletedAt === null))
      .sort((a, b) => new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime());
  }

  async getLatestSnapshot(instanceId: string): Promise<Minion | undefined> {
    const snapshots = await this.listSnapshots(instanceId);
    return snapshots[0];
  }

  async getHistory(instanceId: string): Promise<Minion[]> {
    const storage = await readStorage();
    const snapshotIds = storage.relations
      .filter(r => r.sourceId === instanceId && r.type === 'parent_of')
      .map(r => r.targetId);
    const snapshots = storage.minions.filter(
      m => snapshotIds.includes(m.id) && m.minionTypeId === openclawSnapshotType.id && (m.deletedAt === undefined || m.deletedAt === null)
    );

    // Walk the follows chain to order newest → oldest
    const followsMap = new Map<string, string>();
    for (const r of storage.relations) {
      if (r.type === 'follows' && snapshotIds.includes(r.sourceId)) {
        followsMap.set(r.sourceId, r.targetId);
      }
    }

    // Find the head: the snapshot not referenced as a target in any follows relation
    const targets = new Set(followsMap.values());
    const head = snapshots.find(s => !targets.has(s.id));
    if (!head) return snapshots.sort((a, b) => new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime());

    const ordered: Minion[] = [];
    const byId = new Map(snapshots.map(s => [s.id, s]));
    let current: Minion | undefined = head;
    while (current) {
      ordered.push(current);
      const nextId = followsMap.get(current.id);
      current = nextId ? byId.get(nextId) : undefined;
    }
    return ordered;
  }

  async compare(snapshotId1: string, snapshotId2: string): Promise<Record<string, { from: unknown; to: unknown }>> {
    const storage = await readStorage();
    const a = storage.minions.find(m => m.id === snapshotId1);
    const b = storage.minions.find(m => m.id === snapshotId2);
    if (!a || !b) throw new Error(`Snapshot not found: ${!a ? snapshotId1 : snapshotId2}`);
    return this.diffSnapshots(a, b);
  }

  diffSnapshots(a: Minion, b: Minion): Record<string, { from: unknown; to: unknown }> {
    const diff: Record<string, { from: unknown; to: unknown }> = {};
    const allKeys = new Set([...Object.keys(a.fields), ...Object.keys(b.fields)]);
    for (const key of allKeys) {
      const va = a.fields[key];
      const vb = b.fields[key];
      if (JSON.stringify(va) !== JSON.stringify(vb)) {
        diff[key] = { from: va, to: vb };
      }
    }
    return diff;
  }
}
